### ISSUE TITLE
JDK 17: Files.getFileAttributeView started throwing undeclared UnsupportedOperationException

### ISSUE THREAD

**Issue by mduft** [adoptium/adoptium-support#363]
## Summary

In JDK 8 and JDK 11, Files.getFileAttributeView with an unsupported attribute view returned null, which is also mentioned in the javadoc of this method.

In JDK 17 the javadoc is the same, but a not-mentioned UnsupportedOperationException is thrown.

## Steps to reproduce

Existing code is structured like this:

```
private void setExecutable(Path child, ContentInfo hint) throws IOException {
        PosixFileAttributeView view = Files.getFileAttributeView(child, PosixFileAttributeView.class);
        if (view != null) {
            hint = PathHelper.getContentInfo(child, hint);
            if (hint == null) {
                return;
            }

            if (PathHelper.isExecutable(hint)) {
                Set<PosixFilePermission> perms = view.readAttributes().permissions();
                perms.add(PosixFilePermission.OWNER_EXECUTE);
                perms.add(PosixFilePermission.GROUP_EXECUTE);
                view.setPermissions(perms);
            }
        }
    }
```

This code now leads to an unexpected exception...

## Expected results

Either the API changed and the documentation must be updated, or the change is a mishap and should be corrected.

## Actual results

```
	Caused by: java.lang.UnsupportedOperationException: view <interface java.nio.file.attribute.PosixFileAttributeView> is not supported
		at jdk.zipfs/jdk.nio.zipfs.ZipPath.getFileAttributeView(ZipPath.java:716)
		at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.getFileAttributeView(ZipFileSystemProvider.java:198)
		at java.base/java.nio.file.Files.getFileAttributeView(Files.java:1788)
		at io.bdeploy.bhive.objects.ObjectManager.setExecutable(ObjectManager.java:321)
```

## Triaging info

Java version:

openjdk version "17-beta" 2021-09-14
OpenJDK Runtime Environment Temurin-17+35-202109150246 (build 17-beta+35-202109150246)
OpenJDK 64-Bit Server VM Temurin-17+35-202109150246 (build 17-beta+35-202109150246, mixed mode, sharing)

What is your operating system and platform?

Windows 10 21H1  x86-64

How did you install Java?

Download from https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk17/job/jdk17-windows-x64-hotspot/

Did it work before?

Yes, with previous Java versions, tested only LTS releases 8 and 11.

Did you test with other Java versions?

Yes, with the OpenJDK 17 GA release build, same result.

**Comment by jerboaa** [comment]
@mduft Could you please try this reproducer on JDK 17/11 on Windows and tell me what the output is:

```
$ cat TestPosixAttributeView.java
import java.io.ByteArrayInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.PosixFileAttributeView;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public class TestPosixAttributeView {
	
	private static final String TEST_ZIP_NAME = "test-zip-file.zip";
	private static final String ENTRY_CONTENT = "foo-bar-content";
	private static final String ONLY_ENTRY_NAME = "noperm_entry";

    private void createTestZipFile(Path zpath) throws IOException {
        System.out.println("Creating " + zpath + "...");
		try (ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(zpath.toFile()));
				ByteArrayInputStream fis = new ByteArrayInputStream(ENTRY_CONTENT.getBytes());) {

			ZipEntry zipEntry = new ZipEntry(ONLY_ENTRY_NAME);
			zos.putNextEntry(zipEntry);

			byte[] buffer = new byte[1024];
			int len;
			while ((len = fis.read(buffer)) > 0) {
				zos.write(buffer, 0, len);
			}
			zos.closeEntry();
		}
    }
    
    public void prepare(Path zip) {
    	try {
    		System.out.println(zip.toAbsolutePath().toString());
    		createTestZipFile(zip);
    	} catch (IOException e) {
    		throw new RuntimeException("Test preparation failed!");
    	}
    }
    
    public void runTest(Path zip) {
    	PosixFileAttributeView view = null;
		boolean threwException = false;
		try {
			view = Files.getFileAttributeView(zip, PosixFileAttributeView.class);
		} catch (UnsupportedOperationException e) {
			System.out.println("DEBUG: threw exception: " + e);
			threwException = true;
		}
		String osName = System.getProperty("os.name");
		System.out.println("Debug: os.name == " + osName);
		if (osName.contains("win")) {
			// Windows branch
			if (view == null && !threwException) {
				System.out.println("Test PASSED");
			} else {
				throw new RuntimeException("Test failed!");
			}
		} else {
			if (view == null || threwException) {
				throw new RuntimeException("Test failed!");
			} else {
				System.out.println("Non-Windows Test PASSED!");
			}
		}
    }
    
    public void cleanup(Path zip) {
    	try {
    		Files.deleteIfExists(zip);
    	} catch (IOException e) {
    		// ignore
    	}
    }

	public static void main(String[] args) {
		Path zipFile = Path.of(TEST_ZIP_NAME);
		
		TestPosixAttributeView test = new TestPosixAttributeView();
		test.prepare(zipFile);
		try {
			test.runTest(zipFile);
		} finally {
			test.cleanup(zipFile);
		}
	}

}
$ javac TestPosixAttributeView.java
$ java TestPosixAttributeView
```

**Comment by mduft** [comment]
Some first results - same result on every of the tested JDKs:

```
marku@git-bash: /e/other/jdks/win64 $ ll
total 16
drwxr-xr-x 1 marku 197609    0 Sep 17 07:18 ./
drwxr-xr-x 1 marku 197609    0 Jan 20  2021 ../
drwxr-xr-x 1 marku 197609    0 Jul 21 10:38 jdk-11.0.12+7/
drwxr-xr-x 1 marku 197609    0 Nov 10  2020 jdk-11.0.9.1+1/
drwxr-xr-x 1 marku 197609    0 Oct 23  2020 jdk-15.0.1+9/
drwxr-xr-x 1 marku 197609    0 Sep 15 05:07 jdk-17+35/
-rw-r--r-- 1 marku 197609 2513 Sep 17 07:14 TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.9.1+1/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.9.1+1/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Exception in thread "main" java.lang.RuntimeException: Test failed!
        at TestPosixAttributeView.runTest(TestPosixAttributeView.java:62)
        at TestPosixAttributeView.main(TestPosixAttributeView.java:83)
marku@git-bash: /e/other/jdks/win64 $
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.12+7/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.12+7/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Exception in thread "main" java.lang.RuntimeException: Test failed!
        at TestPosixAttributeView.runTest(TestPosixAttributeView.java:62)
        at TestPosixAttributeView.main(TestPosixAttributeView.java:83)
marku@git-bash: /e/other/jdks/win64 $
marku@git-bash: /e/other/jdks/win64 $ ./jdk-17+35/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-17+35/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Exception in thread "main" java.lang.RuntimeException: Test failed!
        at TestPosixAttributeView.runTest(TestPosixAttributeView.java:62)
        at TestPosixAttributeView.main(TestPosixAttributeView.java:83)
marku@git-bash: /e/other/jdks/win64 $
```

**Comment by mduft** [comment]
I think it does not recognize its in windows due to casing, let me correct that quickly.

**Comment by mduft** [comment]
I changed the `osName.contains("win")` to `osName.contains("Win")` to hit the Windows branch, and now:

```
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.9.1+1/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.9.1+1/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Test PASSED
marku@git-bash: /e/other/jdks/win64 $
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.12+7/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.12+7/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Test PASSED
marku@git-bash: /e/other/jdks/win64 $
marku@git-bash: /e/other/jdks/win64 $
marku@git-bash: /e/other/jdks/win64 $ ./jdk-17+35/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-17+35/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Test PASSED
```

**Comment by mduft** [comment]
Ah - I see why this is passing. In my example it seems I was querying on a ZIP entry, in the test you're still querying the ZIP file itself, not one of it's entries.

**Comment by mduft** [comment]
Slightly modified the test case to query an *entry* of the ZIP instead:  [TestPosixAttributeView.zip](https://github.com/adoptium/adoptium-support/files/7182933/TestPosixAttributeView.zip)

This now yields:

```
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.9.1+1/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.9.1+1/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Test PASSED
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.12+7/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-11.0.12+7/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
Debug: os.name == Windows 10
Test PASSED
marku@git-bash: /e/other/jdks/win64 $ ./jdk-17+35/bin/javac.exe TestPosixAttributeView.java
marku@git-bash: /e/other/jdks/win64 $ ./jdk-17+35/bin/java.exe TestPosixAttributeView
E:\other\jdks\win64\test-zip-file.zip
Creating test-zip-file.zip...
DEBUG: threw exception: java.lang.UnsupportedOperationException: view <interface java.nio.file.attribute.PosixFileAttributeView> is not supported
Debug: os.name == Windows 10
Exception in thread "main" java.lang.RuntimeException: Test failed!
        at TestPosixAttributeView.runTest(TestPosixAttributeView.java:63)
        at TestPosixAttributeView.main(TestPosixAttributeView.java:88)
marku@git-bash: /e/other/jdks/win64 $
```

... which now reproduces the issue perfectly :)

**Comment by jerboaa** [comment]
@mduft Excellent. Thanks for the reproducer. I'll take it upstream.

**Comment by jerboaa** [comment]
https://bugs.openjdk.java.net/browse/JDK-8273935

**Comment by karianna** [comment]
Fixed in 17.0.2
